fingerprintdb_cuda.cu

```diff
@@ -100,101 +143,121 @@ void FingerprintDB::copyToGPU(unsigned int fold_factor)
-                similarity_cutoff));
-
-    auto indices_end = d_results_indices.end();
-    auto scores_end = d_results_scores.end();
-    if(similarity_cutoff > 0) {
-        indices_end = thrust::remove_if(d_results_indices.begin(),
-                d_results_indices.end(), d_results_scores.begin(),
-                thrust::logical_not<bool>());
-        scores_end = thrust::remove(d_results_scores.begin(),
-                d_results_scores.end(), 0);
-    }
-    unsigned int indices_size = std::distance(d_results_indices.begin(),
-            indices_end);
+        const int folded_fp_intsize = m_fp_intsize / m_fold_factor;
+        // Use Tanimoto to score similarity of all compounds to query fingerprint
+        thrust::transform(d_results_indices.begin(), d_results_indices.end(),
+                d_results_scores.begin(),
+                TanimotoFunctor(d_ref_fp, folded_fp_intsize, *(storage->m_priv->d_data),
+                    similarity_cutoff));
+        auto indices_end = d_results_indices.end();
+        auto scores_end = d_results_scores.end();
+        if(similarity_cutoff > 0) {
+            indices_end = thrust::remove_if(d_results_indices.begin(),
+                    d_results_indices.end(), d_results_scores.begin(),
+                    thrust::logical_not<bool>());
+            scores_end = thrust::remove(d_results_scores.begin(),
+                    d_results_scores.end(), 0);
+        }
+        unsigned int indices_size = std::distance(d_results_indices.begin(),
+                indices_end);
 
-    // Sort scores & indices vectors descending on score
-    thrust::sort_by_key(d_results_scores.begin(), scores_end,
-            d_results_indices.begin(), thrust::greater<float>());
+        // Sort scores & indices vectors descending on score
+        thrust::sort_by_key(d_results_scores.begin(), scores_end,
+                d_results_indices.begin(), thrust::greater<float>());
 
-    results_to_consider = std::min(indices_size, return_count*m_fold_factor);
+        int results_to_consider = 0;
+        results_to_consider = std::min(indices_size, return_count*m_fold_factor);
 
-    indices.assign(d_results_indices.begin(), 
-            d_results_indices.begin()+results_to_consider);
+        indices.assign(d_results_indices.begin(), 
+                d_results_indices.begin()+results_to_consider);
 
     } catch(thrust::system_error e) {
-        cerr << "Error!  " << e.what() << endl;
+        qDebug() << "Error!  " << e.what();
     }
 
-
     if(m_fold_factor == 1) { // If we don't fold, we can take exact GPU results
         // Push top return_count results to CPU results vectors to be returned
-        for(unsigned int i=0;i<results_to_consider;i++) {
-            results_smiles.push_back(m_smiles[d_results_indices[i]]);
-            results_ids.push_back(m_ids[d_results_indices[i]]);
+        for(auto index : indices) {
+            int offset_index = index + storage->m_index_offset;
```

Comment:
instead of exposing m_index_offset, can storage have getOffsetIndex(unsigned int index)?

Reply:
Had to add a couple functions to also add the logic required in fingerprint lookup to reverse lookup storage/local_index from index.